torch/distributed/_spmd/data_parallel.py

```diff
@@ -283,8 +284,20 @@ def _gen_partial_strategy(mesh: DeviceMesh) -> PlacementStrategy:
     """
     util function to generate a partial strategy
     """
+    if mesh.device_type == "cuda":
+        # ONLY DeviceMesh CUDA (NCCL backend) by default supports
+        # average reduction, avg reduction is needed depending on
+        # the loss function, for most loss function it should do
+        # gradient averaging. There might be certain cases it should
+        # not do gradient averaging (i.e. sum) but it's pretty rare.
+        reduce_op = c10d.ReduceOp.AVG
+    else:
+        # TODO: we should figure out a way to support avg reduction
+        # for non-NCCL backend (i.e. Gloo)
+        reduce_op = c10d.ReduceOp.SUM
```

Comment:
is there a way to clearly warn users about this difference? Otherwise, it might be surprising that different backend lead to different numerical results. Or maybe, we can just let Gloo crash for now:

https://github.com/pytorch/pytorch/blob/e43918b93a2f595f8a4bbfb9e4d195d3e8746130/torch/csrc/distributed/c10d/ProcessGroupGloo.cpp#L193